graal/src/main/java/io/micronaut/graal/reflect/GraalTypeElementVisitor.java

```diff
@@ -216,85 +225,175 @@ public final void finish(VisitorContext visitorContext) {
+                String path = buildNativeImagePath();
+                String resourcesFile = path + RESOURCE_CONFIG_JSON;
 
-                visitorContext.visitMetaInfFile(propsFile).ifPresent(gf -> {
-                    visitorContext.info("Writing " + NATIVE_IMAGE_PROPERTIES + " file to destination: " + gf.getName());
-                    try (PrintWriter w = new PrintWriter (gf.openWriter())) {
-                        w.println("Args = -H:ReflectionConfigurationResources=${.}/reflection-config.json");
-                    } catch (IOException e) {
-                        e.printStackTrace();
-                    }
-                });
-                final Optional<GeneratedFile> generatedFile = visitorContext.visitMetaInfFile(reflectFile);
+                final Optional<GeneratedFile> generatedFile = visitorContext.visitMetaInfFile(resourcesFile);
                 generatedFile.ifPresent(gf -> {
-                    for (Map<String, Object> value : classes.values()) {
-                        json.add(value);
-                    }
+                    List<Map> resourceList = resourceFiles.stream()
+                            .map(resourceFile -> CollectionUtils.mapOf(PATTERN, "\\Q" + resourceFile + "\\E"))
+                            .collect(Collectors.toList());
 
-                    for (String array : arrays) {
-                        json.add(CollectionUtils.mapOf(
-                                NAME, "[L" + array.substring(0, array.length() - 2) + ";",
-                                ALL_DECLARED_CONSTRUCTORS, true
-                        ));
-                    }
+                    // add any existing resource defined by the user in it's own file in src/main/graal
+                    resourceList.addAll((List) json.get(RESOURCES));
 
-                    try (Writer w = gf.openWriter()) {
-                        visitorContext.info("Writing " + REFLECTION_CONFIG_JSON + " file to destination: " + gf.getName());
+                    json.put(RESOURCES, resourceList);
+                    json.put(BUNDLES, Collections.emptyList());
 
+                    try (Writer w = gf.openWriter()) {
+                        visitorContext.info("Writing " + RESOURCE_CONFIG_JSON + " file to destination: " + gf.getName());
                         writer.writeValue(w, json);
                     } catch (IOException e) {
-                        visitorContext.fail("Error writing " + REFLECTION_CONFIG_JSON + ": " + e.getMessage(), null);
+                        visitorContext.fail("Error writing " + RESOURCE_CONFIG_JSON + ": " + e.getMessage(), null);
                     }
                 });
-            } finally {
-                classes.clear();
-                arrays.clear();
+            } catch (Exception e) {
+                // skip processing resources
+            }
+        }
+    }
+
+    private Set<String> findResourceFiles(File folder, List<String> filePath) {
```

Comment:
Would probably be better implemented using a file tree walker https://docs.oracle.com/javase/tutorial/essential/io/walk.html

But ok

Reply:
I've give it a try and created a FileVisitor but it's not that simple to keep the current directory to prepend it to every file during the visit. As you are ok with this I think I'll keep the current implementation.